File: benchmarks/OCaml/safety/objects/contain.ml
class virtual list = object
  method virtual contain : int -> bool
end

class nil = object
  inherit list
  method contain key = false
end

class cons (hd:int) (tl:list) = object
  inherit list
  method contain key =
    if key = hd then true else tl#contain key
end

let main () =
  assert ((new cons 1 (new cons 2 (new nil)))#contain 2)

File: benchmarks/Java/safety/simple/contain.java
class List extends Object {
    List() {
        super();
    }
    Boolean contain(Int key) {
        assert false;
    }
}

class Nil extends List {
    Nil() {
        super();
    }
    Boolean contain(Int key) {
        return false;
    }
}

class Cons extends List {
    Int hd;
    List tl;
    Cons(Int hd, List tl) {
        super();
        this.hd = hd;
        this.tl = tl;
    }
    Boolean contain(Int key) {
        return if key = hd then true else this.tl.contain(key);
    }
}

main(){
    assert new Cons(1, new Cons(2, new Nil())).contain(2);
}
